Cluster validation improvements

jpanasiukTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/32256

### Problem description
- Cluster validation tool provided limited visibility into physical port configuration, only displaying logical ethernet channels without corresponding physical port IDs or types, making hardware-level debugging difficult
- Workload execution could hang indefinitely without user notification or timeout, leaving validation tool unresponsive and providing no actionable diagnostics
- Missing validation re-checks after detecting failures meant users couldn't determine if hangs were caused by link degradation during workload execution

### What's changed
- Physical port tracking: Extended EthChannelIdentifier with `port_id` and `port_type` fields, implemented `generate_port_info()` to map channels to physical ports via board descriptors, and updated `print_ethernet_connectivity()` and `log_link_metrics()` to display physical port information for hardware-level debugging
- Hang detection and timeout: Converted workload execution to async model with 30-second configurable timeout, using `std::async` with timeout monitoring and distributed all-reduce coordination to ensure all ranks exit together when hangs are detected
- Post-hang validation: Added `handle_workload_timeout()` that re-runs physical discovery and validates current system state against Factory System Descriptor after timeouts, generating timestamped GSD snapshots for post-mortem analysis of link failures during execution
- Code quality improvements: added documentation for ClusterContext and PortInfo structs

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

jpanasiukTT committed:
Cluster validation improvements

- Track and display port_id in print_ethernet_connectivity
- Add port_id and port_type to log_link_metrics output (console and CSV)
- Update get_asic_position to use UMD value for Wormhole instead of discover_tunnels_from_mmio_device
- Update protobuf schema and serialization for new EthChannelIdentifier fields

jpanasiukTT committed:
generate port info move

jpanasiukTT committed:
Add traffic sending hang detection and timeout handling

- Implement async traffic sending to detect hangs during cluster validation
- Add configurable timeout for workload execution (default 30s)
- Improve test loop structure to better track which connections fail
- Add early termination on hang detection to speed up validation

jpanasiukTT committed:
Improve logging and code readability

- Enhance log messages for better debugging
- Refactor code structure for improved maintainability
- Add more descriptive variable names

tt-asaigal commented on tools/scaleout/validation/utils/cluster_validation_utils.cpp:
```diff
@@ -872,24 +816,31 @@ void log_link_metrics(
+        std::cout << std::left << std::setw(9) << row.channel_id.port_id;
+
+        // Print Port Type
+        auto port_type = static_cast<tt::scaleout_tools::PortType>(row.channel_id.port_type);
```
Please use enchantum here for better readability.

jpanasiukTT replied:
What enchantum function would you see used here? This is kinda tricky situation because we have `uint32_t port_type` which we cast to an enum which we want to print thereafter (and we're doing this already using enchantum to_string), all I can think of is using enchantum cast but I don't like the cast because it obfuscates already simple static cast

tt-asaigal replied:
I see it already being used - please disregard this comment.

tt-asaigal commented on tools/scaleout/validation/utils/cluster_validation_utils.cpp:
```diff
@@ -1290,11 +1337,17 @@ bool generate_link_metrics(
         std::map<int, std::shared_ptr<tt::tt_metal::distributed::MeshDevice>> empty_devices;
-        dump_link_stats(inputs, physical_system_descriptor, asic_id_to_chip_id, statuses_per_link, empty_devices, 0, 0);
+        ClusterContext ctx{physical_system_descriptor, asic_id_to_chip_id, empty_devices};
+        dump_link_stats(ctx, inputs, statuses_per_link, 0, 0);
```
I know this is copy pasted code - but for readability can be please do something like this:
`dump_link_stats(ctx, inputs, statuses_per_link, 0 /*data_size*/, 0 /*packet_size_bytes*/);`

tt-asaigal commented on tools/scaleout/validation/utils/cluster_validation_utils.cpp:
```diff
@@ -263,17 +267,37 @@ void execute_workloads(
+            TIMEOUT_DURATION - std::chrono::duration_cast<std::chrono::seconds>(elapsed), std::chrono::seconds(0));
+
+        if (future.wait_for(remaining) != std::future_status::ready) {
+            log_output_rank0("ERROR: Workload execution timed out after 30 seconds");
```
Don't need a print here, since the failure is already logged in `handle_workload_timeout`.

tt-asaigal commented on tools/scaleout/validation/utils/cluster_validation_utils.cpp:
```diff
@@ -965,6 +918,100 @@ void log_link_metrics(
+    log_output_rank0("ERROR: Workload execution timed out after 30 seconds");
+    log_output_rank0("Cluster is not in a healthy state.");
+
+    log_output_rank0("Current ethernet link status:");
```
Don't think we need this log.

tt-asaigal commented on tools/scaleout/validation/utils/cluster_validation_utils.cpp:
```diff
@@ -965,6 +918,100 @@ void log_link_metrics(
+    size_t data_size,
+    size_t packet_size_bytes,
+    bool log_ethernet_metrics) {
+    log_output_rank0("ERROR: Workload execution timed out after 30 seconds");
```
I think we should add the step that validates the Global System Descriptor (live) to Factory System Descriptor to this PR, when a hang is encountered.

Specifically - we currently do this in `run_cluster_validation.cpp::validate_connectivity`.

This will require a minor refactor to the existing code, where `validate_connectivity` takes in a path to the GSD YAML and the FSD textproto.

We can modify the function generating the factory system descriptor as follows:

```
std::string get_factory_system_descriptor_path(
            const std::optional<std::string>& cabling_descriptor_path,
            const std::optional<std::string>& deployment_descriptor_path,
            const std::optional<std::string>& fsd_path,
            const std::string& output_path);
```

We can then move this function to `cluster_validation_utils.cpp`, along with `validate_connectivity`, which also needs a modified signature to remove it's dependence on the input args.

Note that before we validate connectivity upon a workload hang, we must rerun discovery using `physical_system_descriptor.run_discovery(true, true)`.

tt-asaigal commented on tt_metal/fabric/physical_system_descriptor.cpp:
```diff
@@ -83,19 +83,9 @@ std::pair<TrayID, ASICLocation> get_asic_position(
-            // Derive ASIC Location based on the tunnel depth for Wormhole systems
-            // TODO: Remove this once UMD populates the ASIC Location for WH systems.
-            auto mmio_device = cluster_desc->get_closest_mmio_capable_chip(chip_id);
-            auto tunnels_from_mmio_device = llrt::discover_tunnels_from_mmio_device(cluster);
```
With your change, `tt_metal/llrt/tunnels_from_mmio_device.hpp` no longer needs to be included.

tt-asaigal reviewed (commented):
Left a few comments and suggestions for next steps.

Please update the PR description with a sample output, showing how the port id and port type are printed as part of the metrics now.

Please also provide an example of the output a user would see if their validation hung.

Finally, please run Post Commit, T3K & Multi-Host CI and link the passing runs.

jpanasiukTT committed:
Add GSD to FSD validation after hang detection

- Validate Global System Descriptor against Factory System Descriptor after detecting hangs
- Improve error reporting for topology mismatches
- Add better handling of validation failures

tt-asaigal commented on tools/scaleout/validation/utils/cluster_validation_utils.cpp:
```diff
@@ -23,6 +24,10 @@
 
 namespace tt::scaleout_tools {
 
+// Timeout is based on the assumption that 30s should be enough for an iteration to run.
```
valid comment - please address

jpanasiukTT committed:
Code improvemenets

jpanasiukTT committed:
Initialize PortInfo struct members to fix clang-tidy warning

Riddy21 reviewed (approved):
Looks good in terms of the PSD usage and physical discovery stuff. But should double check with Aditya on the rest

jpanasiukTT committed:
update comment

jpanasiukTT committed:
remove redundant {} initializers

jpanasiukTT committed:
fix clang tidy
